### PR TITLE
251 custom decorator fix product test

### PR DESCRIPTION
**Pull request by leungkinghin**
#### Summary
Resolve test case for Product Resources

#### Description
After client decoration in next featured branch https://github.com/commercetools/commercetools-sync-java/commit/e57c36407487c0813bb3761b170df244c8628e83, we have to resolve test cases failure in IT and UT. The first step is to implement different resolutions for product test cases.

#### Relevant Issues
#251 

#### Todo

- Tests
    - [x] Unit 
    - [x] Integration
- [ ] Documentation
- [ ] Add Release Notes entry.

### CODE REVIEW

File: src/test/java/com/commercetools/sync/services/impl/ProductServiceTest.java [hunk 1]
```diff
@@ -149,4 +153,96 @@ void buildProductKeysQueryPredicate_WithSomeBlankProductKeys_ShouldBuildCorrectQ
...
+                .errorCallback((exception, oldResource, newResource, updateActions) -> {
+                    errorMessages.add(exception.getMessage());
+                    errorExceptions.add(exception.getCause());
+                })
+                .build();
+
+        service = new ProductServiceImpl(productSyncOptions);
+    }
+
+    private static class FakeProductClient implements SphereClient {
```
Comment (ahmetoz) [inline src/test/java/com/commercetools/sync/services/impl/ProductServiceTest.java:208]: Probably you will need this for other service test, what do you think to make it generic ?

File: src/test/java/com/commercetools/sync/services/impl/ProductServiceTest.java [hunk 2]
```diff
@@ -149,4 +153,96 @@ void buildProductKeysQueryPredicate_WithSomeBlankProductKeys_ShouldBuildCorrectQ
...
+        final FakeProductClient fakeProductClient = new FakeProductClient(new BadGatewayException());
+
+        initMockService(fakeProductClient);
+
+        CompletionStage<Optional<Product>> future = service.fetchProduct("productKey1");
+
+        assertThat(errorExceptions).isEmpty();
+        assertThat(errorMessages).isEmpty();
+        assertThat(future.toCompletableFuture())
+                .failsWithin(60, TimeUnit.SECONDS)
```
Comment (ahmetoz) [inline src/test/java/com/commercetools/sync/services/impl/ProductServiceTest.java:187]: Why you need 60 seconds ?

File: src/test/java/com/commercetools/sync/services/impl/ProductServiceTest.java [hunk 3]
```diff
@@ -149,4 +153,96 @@ void buildProductKeysQueryPredicate_WithSomeBlankProductKeys_ShouldBuildCorrectQ
...
+                .withThrowableOfType(ExecutionException.class)
+                .withCauseExactlyInstanceOf(BadGatewayException.class);
+
+    }
+
+    @Nonnull
```
Comment (ahmetoz) [inline src/test/java/com/commercetools/sync/services/impl/ProductServiceTest.java:193]: `Nonnull` annotation is unncessary.

**Review by ahmetoz — CHANGES_REQUESTED** [review]
It would be nice to know coverage change after this update, but hard to reason because the build is broken 😢

File: src/test/java/com/commercetools/sync/services/impl/ProductServiceTest.java (hunk 2 above)
Comment (leungkinghin) [reply src/test/java/com/commercetools/sync/services/impl/ProductServiceTest.java:187]: It is derived from a deprecated method `hasFailedWithThrowableThat()`
Javadoc suggests to use `failsWithin()` instead. It needs a timeout limit of service class async call, hence i put 60 seconds as default value here.

File: src/test/java/com/commercetools/sync/services/impl/ProductServiceTest.java (hunk 1 above)
Comment (leungkinghin) [reply src/test/java/com/commercetools/sync/services/impl/ProductServiceTest.java]: https://github.com/commercetools/commercetools-sync-java/pull/622/commits/47c56027a46c4d4b66b3c865fbace19b7d892fad

File: src/test/java/com/commercetools/sync/services/impl/ProductServiceTest.java (hunk 3 above)
Comment (leungkinghin) [reply src/test/java/com/commercetools/sync/services/impl/ProductServiceTest.java]: https://github.com/commercetools/commercetools-sync-java/pull/622/commits/4b68a84f3fd74b502354698b9d4005df30faf73c

File: src/test/java/com/commercetools/sync/services/impl/ProductServiceTest.java (hunk 2 above)
Comment (ahmetoz) [reply src/test/java/com/commercetools/sync/services/impl/ProductServiceTest.java:185]: OK, Javadocs of the hasFailedWithThrowableThat is suggesting 1 second.  Don't know where you got 60seconds default? Anyways you might skip my comment.

````
Although not 100% the same, consider using {@link #failsWithin(Duration)} or {@link #failsWithin(long, TimeUnit)} instead:`
   
  <pre><code class='java'> CompletableFuture future = new CompletableFuture();
  future.completeExceptionally(new RuntimeException("boom!"));
   
    assertThat(future).failsWithin(1, TimeUnit.SECONDS)
                      .withThrowableOfType(RuntimeException.class)
                      .withMessage("boom!");
````

**Review by ahmetoz — APPROVED** [review]
Approved only for product test changes, maybe you need to open a different PR for others.

File: src/test/java/com/commercetools/sync/services/impl/ProductServiceTest.java (hunk 2 above)
Comment (leungkinghin) [reply src/test/java/com/commercetools/sync/services/impl/ProductServiceTest.java]: Choose it randomly.

File: src/test/java/com/commercetools/sync/commons/FakeClient.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+    @Override
+    public <T> CompletionStage<T> execute(final SphereRequest<T> sphereRequest) {
+        isExecuted = true;
+        if (mockResult != null) {
+            return CompletableFutureUtils.successful((T)mockResult);
+        } else if (mockException != null) {
+            return CompletableFutureUtils.failed(mockException);
+        } else {
+            return CompletableFutureUtils.successful(null);
```
Comment (lojzatran) [inline src/test/java/com/commercetools/sync/commons/FakeClient.java:33]: This part of if branch is never reachable, you can remove it.

File: src/test/java/com/commercetools/sync/products/ProductSyncTest.java [hunk 5]
```diff
@@ -356,13 +364,263 @@ void sync_WithEmptyAttributeMetaDataMap_ShouldCallErrorCallback() {
...
+                mock(CustomObjectService.class));
+
+        final ProductSyncStatistics syncStatistics = productSync
+                .sync(singletonList(productDraft))
+                .toCompletableFuture()
+                .join();
+
+        // assertion
+
+        // Test and assertion
```
Comment (lojzatran) [inline src/test/java/com/commercetools/sync/products/ProductSyncTest.java:466]: Why is this part called `// Test and assertion`?

File: src/test/java/com/commercetools/sync/services/impl/ProductServiceTest.java [hunk 6]
```diff
@@ -42,31 +43,28 @@
...
     }
 
     @Test
     void createProduct_WithSuccessfulMockCtpResponse_ShouldReturnMock() {
         final Product mock = mock(Product.class);
         when(mock.getId()).thenReturn("productId");
         when(mock.getKey()).thenReturn("productKey");
-
-        when(productSyncOptions.getCtpClient().execute(any())).thenReturn(completedFuture(mock));
+        final FakeClient createProductClient =
```
Comment (lojzatran) [inline src/test/java/com/commercetools/sync/services/impl/ProductServiceTest.java:55]:
```suggestion
        final FakeClient<Product> createProductClient =
```

File: src/test/java/com/commercetools/sync/services/impl/ProductServiceTest.java [hunk 7]
```diff
@@ -42,31 +43,28 @@
...
-        productSyncOptions = ProductSyncOptionsBuilder
-            .of(mock(SphereClient.class))
-            .errorCallback((exception, oldResource, newResource, updateActions) -> {
-                errorMessages.add(exception.getMessage());
-                errorExceptions.add(exception.getCause());
-            })
-            .build();
-        service = new ProductServiceImpl(productSyncOptions);
+        initMockService(mock(SphereClient.class));
+
```
Comment (lojzatran) [inline src/test/java/com/commercetools/sync/services/impl/ProductServiceTest.java:47]: Remove empty line

File: src/test/java/com/commercetools/sync/services/impl/ProductServiceTest.java [hunk 8]
```diff
@@ -149,4 +151,54 @@ void buildProductKeysQueryPredicate_WithSomeBlankProductKeys_ShouldBuildCorrectQ
         final QueryPredicate<Product> queryPredicate = service.buildProductKeysQueryPredicate(productKeys);
         assertThat(queryPredicate.toSphereQuery()).isEqualTo("key in (\"key1\", \"key2\")");
     }
+
+    @Test
+    void fetchMatchingProductsByKeys_WithBadGateWayExceptionAlways_ShouldFail() {
```
Comment (lojzatran) [inline src/test/java/com/commercetools/sync/services/impl/ProductServiceTest.java:156]:
```suggestion
    void fetchMatchingProductsByKeys_WithBadGateWayException_ShouldFail() {
```

**Review by lojzatran — CHANGES_REQUESTED** [review]

File: src/test/java/com/commercetools/sync/commons/FakeClient.java (hunk 4 above)
Comment (leungkinghin) [reply src/test/java/com/commercetools/sync/commons/FakeClient.java]: https://github.com/commercetools/commercetools-sync-java/pull/622/commits/8b1a907d7a65a2707752e4073fdac629715c7319

File: src/test/java/com/commercetools/sync/products/ProductSyncTest.java (hunk 5 above)
Comment (leungkinghin) [reply src/test/java/com/commercetools/sync/products/ProductSyncTest.java]: https://github.com/commercetools/commercetools-sync-java/pull/622/commits/685236a17bdb6c7e2af5d34da44a37ddfcac47a1

File: src/test/java/com/commercetools/sync/commons/FakeClient.java (hunk 4 above)
Comment (lojzatran) [reply src/test/java/com/commercetools/sync/commons/FakeClient.java]:
```
        if (mockResult != null) {
            return CompletableFutureUtils.successful((T)mockResult);
        } else {
            return CompletableFutureUtils.failed(mockException);
        }
```

File: src/test/java/com/commercetools/sync/services/impl/ProductServiceTest.java (hunk 6 above)
Comment (leungkinghin) [reply src/test/java/com/commercetools/sync/services/impl/ProductServiceTest.java:55]: https://github.com/commercetools/commercetools-sync-java/pull/622/commits/1d59d25807d16f99dc0da8f4ac79d4de313a487b

File: src/test/java/com/commercetools/sync/services/impl/ProductServiceTest.java (hunk 8 above)
Comment (leungkinghin) [reply src/test/java/com/commercetools/sync/services/impl/ProductServiceTest.java]: https://github.com/commercetools/commercetools-sync-java/pull/622

File: src/test/java/com/commercetools/sync/services/impl/ProductServiceTest.java (hunk 7 above)
Comment (leungkinghin) [reply src/test/java/com/commercetools/sync/services/impl/ProductServiceTest.java]: https://github.com/commercetools/commercetools-sync-java/pull/622/commits/9904f4b4d54fa222f211a0ef1a93a1a8bb0fada2

File: src/test/java/com/commercetools/sync/commons/FakeClient.java (hunk 4 above)
Comment (leungkinghin) [reply src/test/java/com/commercetools/sync/commons/FakeClient.java]: https://github.com/commercetools/commercetools-sync-java/pull/622/commits/9d153ac27794cffc6c1ed8a685c69fe300919f08

**Review by lojzatran — APPROVED** [review]

**Pull request merged**